src/pms/core/reader.py

```diff
@@ -119,7 +162,29 @@ def _pre_heat(self):
         # only pre-heat the firs time
         self.pre_heat = 0
 
-    def __enter__(self) -> "SensorReader":
+    def _read_one(self) -> _Reading:
+        """Return a single passive mode reading"""
+
+        if not self.serial.is_open:
+            raise StopIteration
+
+        buffer = self._cmd("passive_read")
+
+        try:
+            obs = self.sensor.decode(buffer)
+            return self._Reading(buffer=buffer, obs_data=obs)
+        except SensorNotReady as e:
+            # no special hardware handling
+            raise
+        except SensorWarning as e:  # pragma: no cover
+            self.serial.reset_input_buffer()
+            raise
```

Comment:
Please correct me if I'm wrong.
The reason for this method is to read one observation, without going trough `__call__`.
From your [comment on PR 32](https://github.com/avaldebe/PyPMS/pull/32#issuecomment-1328059992), I understand that you want to avoid `__call__` because it will trigger a delay/sleep before the next sample.

I think that setting `samples = 1` or `interval = None` will bypass the delay/sleep.
What do you think of this approach?

Reply:
on https://github.com/avaldebe/PyPMS/compare/drop-python-3.7...read-one I checked out your tests and docs from this PR and implemented the methods discussed on #32
It passes your tests and runs your example (with some minor updates).
Would this be good enough for your usage?

Reply:
I'm afraid not, sorry.

The problem is the error handling in `__call__`, which is written for CLI usage, not library usage:

- [It swallows errors](https://github.com/benthorner/PyPMS/blob/f8d6d59a514aca757c0f8dd7d0f5b4b9b84fa2d5/src/pms/core/reader.py#L168-L170), which means the calling code could block for a long time.
- It forces the calling code to block for certain types of error ([the `sleep(5)`](https://github.com/benthorner/PyPMS/blob/f8d6d59a514aca757c0f8dd7d0f5b4b9b84fa2d5/src/pms/core/reader.py#L167) line).

You could fix this with more flags on the class or the function, but I think that'll just make the class harder and harder to understand; whereas `read_one` exposes all the internals directly. Although it may be subjective, to me it also makes sense that `__call__` uses `(_)read_one`, as the abstraction actually helps to follow what it does: 

1. (new) "`__call__` _reads one_ sample at a time, in intervals, until it has enough to return" vs.
2. (old) "`__call__` sends a passive read command and decodes the data, in intervals, until it has enough to return"

You and I both know that (2) is really the same as (1), but to someone new trying to contribute to this code (or maybe even you coming back to it in a year's time!) it's harder to see this is the case unless you "just know".

What do you see as the drawbacks of `(_)read_one` doing the heavy lifting?

Reply:
@benthorner can you provide a test that exemplifies the problem?

Reply:
Sure, I've added an extra commit with a couple of tests that will only pass with the approach here.

The behaviour I need is for `read_one` to _try_ to read, not _force_ a read. I've updated the PR code snippet to clarify: for a single poll, it makes more sense to give up ("fail") if there's a problem, rather than force the whole system wait until a reading is available; by definition, the system will try again soon anyway, so there's no point waiting.

This is the difference between the approach here, which can fail instead of returning data, and the use of `__call__`, which will try to force a reading no matter how long it takes using its own, non-customisable `sleep`.

Does that help?